[no-unused-vars] v4 false positive with generic inference - with isolated reproduction repo

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"

Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Are you opening an issue because the rule you're trying to use is not found?
🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨
1) Check the releases log: https://github.com/typescript-eslint/typescript-eslint/releases
    -  If the rule isn't listed there, then chances are it hasn't been released to the main npm tag yet.
2) Try installing the `canary` tag: `npm i @typescript-eslint/eslint-plugin@canary`.
    - The canary tag is built for every commit to master, so it contains the bleeding edge build.
3) If ESLint still can't find the rule, then consider reporting an issue.
-->

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
https://github.com/willheslam/typescript-unused-var-reprod
`npm run lint`

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.

Please consider creating an isolated reproduction repo to make it easy for the volunteer maintainers debug your issue.
-->

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": 1
  }
}
```

```TS
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never
```

<!--
Also include your tsconfig, if you're using type-aware linting
-->

**Expected Result**

No linting issues

<!--
What did you expect to happen?
Please be specific here - list the exact lines and messages you expect.
-->

**Actual Result**

```
3:12  warning  'I' is defined but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (0 errors, 1 warning)
```

<!--
What actually happened?
Please be specific here - list the exact lines and messages that caused errors
-->

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.0` |
| `node`                             | `12.14.1` |

I'm seeing a similar warning in this code:

```
// https://stackoverflow.com/a/51955852/
type KnownKeys<T> = {
    [K in keyof T]: string extends K ? never : number extends K ? never : K
} extends { [_ in keyof T]: infer U } ? U : never;
```

gives

```
  4:35  error  'U' is defined but never used  @typescript-eslint/no-unused-vars
```

--- a/packages/eslint-plugin/tests/rules/no-unused-vars.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-unused-vars.test.ts
@@ -789,6 +789,20 @@ declare const [v6];
       `,
       filename: 'foo.d.ts',
     },
+    // https://github.com/typescript-eslint/typescript-eslint/issues/2459
+    `
+export type Test<U> = U extends (k: infer I) => void ? I : never;
+    `,
+    `
+export type Test<U> = U extends { [k: string]: infer I } ? I : never;
+    `,
+    `
+export type Test<U> = U extends (arg: {
+  [k: string]: (arg2: infer I) => void;
+}) => void
+  ? I
+  : never;
+    `,
   ],
 
   invalid: [

--- a/packages/scope-manager/src/referencer/TypeVisitor.ts
+++ b/packages/scope-manager/src/referencer/TypeVisitor.ts
@@ -2,6 +2,7 @@ import { TSESTree, AST_NODE_TYPES } from '@typescript-eslint/types';
 import { Referencer } from './Referencer';
 import { Visitor } from './Visitor';
 import { ParameterDefinition, TypeDefinition } from '../definition';
+import { ScopeType } from '../scope';
 
 class TypeVisitor extends Visitor {
   readonly #referencer: Referencer;
@@ -119,6 +120,44 @@ class TypeVisitor extends Visitor {
       }
     }
     this.visit(node.typeAnnotation);
+  }
+
+  protected TSInferType(node: TSESTree.TSInferType): void {
+    const typeParameter = node.typeParameter;
+    let scope = this.#referencer.currentScope();
+
+    /*
+    In cases where there is a sub-type scope created within a conditional type, then the generic should be defined in the
+    conditional type's scope, not the child type scope.
+    If we define it within the child type's scope then it won't be able to be referenced outside the child type
+    */
+    if (
+      scope.type === ScopeType.functionType ||
+      scope.type === ScopeType.mappedType
+    ) {
+      // search up the scope tree to figure out if we're in a nested type scope
+      let currentScope = scope.upper;
+      while (currentScope) {
+        if (
+          currentScope.type === ScopeType.functionType ||
+          currentScope.type === ScopeType.mappedType
+        ) {
+          // ensure valid type parents only
+          currentScope = currentScope.upper;
+          continue;
+        }
+        if (currentScope.type === ScopeType.conditionalType) {
+          scope = currentScope;
+          break;
+        }
+        break;
+      }
+    }
+
+    scope.defineIdentifier(
+      typeParameter.name,
+      new TypeDefinition(typeParameter.name, typeParameter),
+    );
   }
 
   protected TSInterfaceDeclaration(

--- a/packages/scope-manager/tests/fixtures/type-declaration/conditional3.ts
+++ b/packages/scope-manager/tests/fixtures/type-declaration/conditional3.ts
@@ -1,0 +1,1 @@
+type Test<U> = U extends (k: infer I) => void ? I : never;

--- a/packages/scope-manager/tests/fixtures/type-declaration/conditional3.ts.shot
+++ b/packages/scope-manager/tests/fixtures/type-declaration/conditional3.ts.shot
@@ -1,0 +1,131 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`type-declaration conditional3 1`] = `
+ScopeManager {
+  variables: Array [
+    Variable$1 {
+      defs: Array [
+        TypeDefinition$1 {
+          name: Identifier<"Test">,
+          node: TSTypeAliasDeclaration$1,
+        },
+      ],
+      name: "Test",
+      references: Array [],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+    Variable$2 {
+      defs: Array [
+        TypeDefinition$2 {
+          name: Identifier<"U">,
+          node: TSTypeParameter$2,
+        },
+      ],
+      name: "U",
+      references: Array [
+        Reference$1 {
+          identifier: Identifier<"U">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$2,
+        },
+      ],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+    Variable$3 {
+      defs: Array [
+        ParameterDefinition$3 {
+          name: Identifier<"k">,
+          node: TSFunctionType$3,
+        },
+      ],
+      name: "k",
+      references: Array [],
+      isValueVariable: true,
+      isTypeVariable: false,
+    },
+    Variable$4 {
+      defs: Array [
+        TypeDefinition$4 {
+          name: Identifier<"I">,
+          node: TSTypeParameter$4,
+        },
+      ],
+      name: "I",
+      references: Array [
+        Reference$2 {
+          identifier: Identifier<"I">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$4,
+        },
+      ],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+  ],
+  scopes: Array [
+    GlobalScope$1 {
+      block: Program$5,
+      isStrict: false,
+      references: Array [],
+      set: Map {
+        "Test" => Variable$1,
+      },
+      type: "global",
+      upper: null,
+      variables: Array [
+        Variable$1,
+      ],
+    },
+    TypeScope$2 {
+      block: TSTypeAliasDeclaration$1,
+      isStrict: true,
+      references: Array [],
+      set: Map {
+        "U" => Variable$2,
+      },
+      type: "type",
+      upper: GlobalScope$1,
+      variables: Array [
+        Variable$2,
+      ],
+    },
+    ConditionalTypeScope$3 {
+      block: TSConditionalType$6,
+      isStrict: true,
+      references: Array [
+        Reference$1,
+        Reference$2,
+      ],
+      set: Map {
+        "I" => Variable$4,
+      },
+      type: "conditionalType",
+      upper: TypeScope$2,
+      variables: Array [
+        Variable$4,
+      ],
+    },
+    FunctionTypeScope$4 {
+      block: TSFunctionType$3,
+      isStrict: true,
+      references: Array [],
+      set: Map {
+        "k" => Variable$3,
+      },
+      type: "functionType",
+      upper: ConditionalTypeScope$3,
+      variables: Array [
+        Variable$3,
+      ],
+    },
+  ],
+}
+`;

--- a/packages/scope-manager/tests/fixtures/type-declaration/conditional4.ts
+++ b/packages/scope-manager/tests/fixtures/type-declaration/conditional4.ts
@@ -1,0 +1,1 @@
+type Test<U> = U extends { [k: string]: infer I } ? I : never;

--- a/packages/scope-manager/tests/fixtures/type-declaration/conditional4.ts.shot
+++ b/packages/scope-manager/tests/fixtures/type-declaration/conditional4.ts.shot
@@ -1,0 +1,106 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`type-declaration conditional4 1`] = `
+ScopeManager {
+  variables: Array [
+    Variable$1 {
+      defs: Array [
+        TypeDefinition$1 {
+          name: Identifier<"Test">,
+          node: TSTypeAliasDeclaration$1,
+        },
+      ],
+      name: "Test",
+      references: Array [],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+    Variable$2 {
+      defs: Array [
+        TypeDefinition$2 {
+          name: Identifier<"U">,
+          node: TSTypeParameter$2,
+        },
+      ],
+      name: "U",
+      references: Array [
+        Reference$1 {
+          identifier: Identifier<"U">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$2,
+        },
+      ],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+    Variable$3 {
+      defs: Array [
+        TypeDefinition$3 {
+          name: Identifier<"I">,
+          node: TSTypeParameter$3,
+        },
+      ],
+      name: "I",
+      references: Array [
+        Reference$2 {
+          identifier: Identifier<"I">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$3,
+        },
+      ],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+  ],
+  scopes: Array [
+    GlobalScope$1 {
+      block: Program$4,
+      isStrict: false,
+      references: Array [],
+      set: Map {
+        "Test" => Variable$1,
+      },
+      type: "global",
+      upper: null,
+      variables: Array [
+        Variable$1,
+      ],
+    },
+    TypeScope$2 {
+      block: TSTypeAliasDeclaration$1,
+      isStrict: true,
+      references: Array [],
+      set: Map {
+        "U" => Variable$2,
+      },
+      type: "type",
+      upper: GlobalScope$1,
+      variables: Array [
+        Variable$2,
+      ],
+    },
+    ConditionalTypeScope$3 {
+      block: TSConditionalType$5,
+      isStrict: true,
+      references: Array [
+        Reference$1,
+        Reference$2,
+      ],
+      set: Map {
+        "I" => Variable$3,
+      },
+      type: "conditionalType",
+      upper: TypeScope$2,
+      variables: Array [
+        Variable$3,
+      ],
+    },
+  ],
+}
+`;

--- a/packages/scope-manager/tests/fixtures/type-declaration/conditional5.ts
+++ b/packages/scope-manager/tests/fixtures/type-declaration/conditional5.ts
@@ -1,0 +1,3 @@
+type Test<U> = U extends (arg: { [k: string]: (arg2: infer I) => void }) => void
+  ? I
+  : never;

--- a/packages/scope-manager/tests/fixtures/type-declaration/conditional5.ts.shot
+++ b/packages/scope-manager/tests/fixtures/type-declaration/conditional5.ts.shot
@@ -1,0 +1,156 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`type-declaration conditional5 1`] = `
+ScopeManager {
+  variables: Array [
+    Variable$1 {
+      defs: Array [
+        TypeDefinition$1 {
+          name: Identifier<"Test">,
+          node: TSTypeAliasDeclaration$1,
+        },
+      ],
+      name: "Test",
+      references: Array [],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+    Variable$2 {
+      defs: Array [
+        TypeDefinition$2 {
+          name: Identifier<"U">,
+          node: TSTypeParameter$2,
+        },
+      ],
+      name: "U",
+      references: Array [
+        Reference$1 {
+          identifier: Identifier<"U">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$2,
+        },
+      ],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+    Variable$3 {
+      defs: Array [
+        ParameterDefinition$3 {
+          name: Identifier<"arg">,
+          node: TSFunctionType$3,
+        },
+      ],
+      name: "arg",
+      references: Array [],
+      isValueVariable: true,
+      isTypeVariable: false,
+    },
+    Variable$4 {
+      defs: Array [
+        ParameterDefinition$4 {
+          name: Identifier<"arg2">,
+          node: TSFunctionType$4,
+        },
+      ],
+      name: "arg2",
+      references: Array [],
+      isValueVariable: true,
+      isTypeVariable: false,
+    },
+    Variable$5 {
+      defs: Array [
+        TypeDefinition$5 {
+          name: Identifier<"I">,
+          node: TSTypeParameter$5,
+        },
+      ],
+      name: "I",
+      references: Array [
+        Reference$2 {
+          identifier: Identifier<"I">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$5,
+        },
+      ],
+      isValueVariable: false,
+      isTypeVariable: true,
+    },
+  ],
+  scopes: Array [
+    GlobalScope$1 {
+      block: Program$6,
+      isStrict: false,
+      references: Array [],
+      set: Map {
+        "Test" => Variable$1,
+      },
+      type: "global",
+      upper: null,
+      variables: Array [
+        Variable$1,
+      ],
+    },
+    TypeScope$2 {
+      block: TSTypeAliasDeclaration$1,
+      isStrict: true,
+      references: Array [],
+      set: Map {
+        "U" => Variable$2,
+      },
+      type: "type",
+      upper: GlobalScope$1,
+      variables: Array [
+        Variable$2,
+      ],
+    },
+    ConditionalTypeScope$3 {
+      block: TSConditionalType$7,
+      isStrict: true,
+      references: Array [
+        Reference$1,
+        Reference$2,
+      ],
+      set: Map {
+        "I" => Variable$5,
+      },
+      type: "conditionalType",
+      upper: TypeScope$2,
+      variables: Array [
+        Variable$5,
+      ],
+    },
+    FunctionTypeScope$4 {
+      block: TSFunctionType$3,
+      isStrict: true,
+      references: Array [],
+      set: Map {
+        "arg" => Variable$3,
+      },
+      type: "functionType",
+      upper: ConditionalTypeScope$3,
+      variables: Array [
+        Variable$3,
+      ],
+    },
+    FunctionTypeScope$5 {
+      block: TSFunctionType$4,
+      isStrict: true,
+      references: Array [],
+      set: Map {
+        "arg2" => Variable$4,
+      },
+      type: "functionType",
+      upper: FunctionTypeScope$4,
+      variables: Array [
+        Variable$4,
+      ],
+    },
+  ],
+}
+`;

